test/utils.js

```diff
@@ -0,0 +1,41 @@
+goog.module('measurementLibrary.measure.testing.utils');
+goog.setTestOnly();
+
+const setupMeasure = goog.require('measurementLibrary.setup');
+
+const noop = () => {};
+
+/**
+ * Run a test, once assuming that the measure snippet fired before the call
+ * to setup, then again assuming that the measure snippet fired after the call
+ * to setup.
+ * @param {function(!Array<*>):undefined} config The configuration
+ *     to be done in the measure code snippet.
+ * @param {function(!Array<*>):undefined} test The tests to run
+ *     after both the code snippet and setupMeasure function have fired.
+ * @param {function():undefined} reset A function to reset state between tests.
+ */
+function runInBothOrders(config, test, reset = noop) {
```

Comment:
I'm not sure about this pattern. Considering race conditions is commendable, but there are a number of potential issues here:

1. There's a lot of indirection, which can obscure what the test is actually doing. At a minimum, new developers/users of the code will need to switch back and forth between this helper file and the actual test code to understand what's going on. If this will be only used for setup unit tests, I'd move the helper function to that file. I'd strongly recommend against using this plan for all tests.

2. Jasmine has well-accepted ways of resetting state (beforeAll/beforeEach), and this jumps out of that usual flow. 

3. I'm not sure it will be clear to the user exactly where the test fails. Wrapping each separate case in a "describe" or "it" statement could help (I think Jasmine supports multiple nested describes, but not sure about its).

4. IF we decide to keep this method of testing (which I'd recommend against), I think we need a more descriptive name of the helper function.

5. On the topic of testing for race conditions in general, I'd consider whether or not every case merits a unit test: 
- Are there particular features potentially sensitive to race conditions? Those might be worth testing in a unit test.
- Is there something about the library as a whole that is sensitive to race conditions? In that case, it should probably be addressed in integration tests. Yes, the snippet of code which is responsible for ensuring correct behavior when setup/snippet order changes should be covered in a handful of unit tests, but we definitely don't need to do that in every test.



In general, I'd prefer to think of tests as documentation for how the code should work. If a new user wants to understand how the library works, how difficult will it be for them to learn about it from the tests alone? If tests are documentation, then generally they should be more descriptive. This is a somewhat controversial topic in the developer world at large, but Google tends to prefer it this way: https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html

Reply:
Discussed in person
1, 5) Moved the function to just the setup unit tests.
2, 3) Apparently you can nest if blocks inside of functions nested inside of describe blocks!